lvsfunc/source.py

```diff
@@ -0,0 +1,206 @@
+        else:
+            clip = core.ffms2.Source(file, **index_args)
+
+    if ref:
+        if ref.format is None:
+            raise ValueError("source: 'Variable-format clips not supported.'")
+        clip = core.std.AssumeFPS(clip, fpsnum=ref.fps.numerator, fpsden=ref.fps.denominator)
+        clip = core.resize.Bicubic(clip, width=ref.width, height=ref.height,
+                                   format=ref.format.id, matrix=get_matrix(ref))
+        if is_image(file):
+            clip = clip * (ref.num_frames - 1)
+
+    return clip
+
+
+class DVDIndexer(ABC):
+    """Abstract DVD indexer interface."""
+    path: Union[Path, str]
+    vps_indexer: Callable[..., vs.VideoNode]
+    ext: str
+
+    def __init__(self, path: Union[Path, str], vps_indexer: Callable[..., vs.VideoNode], ext: str) -> None:
+        self.path = path
+        self.vps_indexer = vps_indexer  # type: ignore
+        self.ext = ext
+        super().__init__()
+
+    @abstractmethod
+    def get_cmd(self, files: List[Path], output: Path) -> List[Any]:
+        """Returns the indexer command"""
+        raise NotImplementedError
+
+
+class D2VWitch(DVDIndexer):
+    """Built-in d2vwitch indexer"""
+    def __init__(self, path: Union[Path, str] = 'd2vwitch',
+                 vps_indexer: Callable[..., vs.VideoNode] = core.d2v.Source, ext: str = '.d2v') -> None:
+        super().__init__(path, vps_indexer, ext)
+
+    def get_cmd(self, files: List[Path], output: Path) -> List[Any]:
+        return [self.path, '--output', output, *files]
+
+
+class DGIndexNV(DVDIndexer):
+    """Built-in DGIndexNV indexer"""
+    def __init__(self, path: Union[Path, str] = 'DGIndexNV',
+                 vps_indexer: Callable[..., vs.VideoNode] = core.dgdecodenv.DGSource, ext: str = '.dgi') -> None:
+        super().__init__(path, vps_indexer, ext)
+
+    def get_cmd(self, files: List[Path], output: Path) -> List[Any]:
+        return [self.path, '-i', ','.join(map(str, files)), '-o', output, '-h']
+
+
+def dvd_source(vob_folder: Union[Path, str], idx: DVDIndexer = D2VWitch(), ifo_file: Optional[Union[Path, str]] = None,
+               extra: bool = False, **kwargs: Any) -> Tuple[List[vs.VideoNode], List[List[int]]]:
+    """Concat .vob DVD files and trim them per titles using the .ifo file as reference.
+
+    :param vob_folder:      Folder with .vob files.
+    :param idx:             DVD indexer to be used.
+    :param ifo_file:        Optional IFO file path. If not specified, will search 'VTS_01_0.IFO'.
+    :param extra:           Adds the splash screen and BD Menu to the return clips.
+    :kwargs:                Arguments passed to the indexing filter.
+    :return:                Returns a tuple of trimmed clips per title and the chapter frame numbers.
+    """
+    vob_folder = Path(vob_folder)
+
+    # Index vob files using idx
+    vob_files = sorted(vob_folder.glob('*.vob'))
+
+    if not (output := Path(vob_files[0].with_suffix(idx.ext))).exists():
```

Comment:
This currently throws an unhelpful error when no VOB files are found. Would suggest adding an error like this, or adding a way to have it index ISO files directly (d2vwitch at least should be capable of doing this):
```py
    if not vob_files:
        iso_files = vob_folder.glob('*.iso')
        if iso_files:
            raise ValueError("dvd_source: No VOBs found. Following DVDISOs were found. " +
                             "Please unpack the following DVDISOs " +
                             "and point `vob_folder` to your new directories:\n\n" +
                             "\n".join([str(f) for f in iso_files]))
        raise ValueError(f"dvd_source: No VOB or ISO file found in given directory {vob_folder}")
```

Reply:
Fix in https://github.com/Irrational-Encoding-Wizardry/lvsfunc/pull/59/commits/ef1e4baac246a50cc49430874af10ea5053c5bc3.
ISO files ISOs will not be supported since there are no way to get IFO data from the ISO file itself.